clash-cores/src/Clash/Cores/Sgmii/RateAdapt.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE CPP #-}
+
+{- |
+  Copyright   :  (C) 2024, QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Functions for the rate adaptation blocks that are required for lower bit
+  rates than 1000 Mbps
+-}
+module Clash.Cores.Sgmii.RateAdapt
+  ( rateAdaptRx
+  , rateAdaptTx
+  )
+where
+
+import Clash.Cores.Sgmii.Common
+import Clash.Prelude
+
+-- | State transition function for the receive rate adaption function
+rateAdaptRxT ::
+  -- | Current state
+  Index 100 ->
+  -- | Input value
+  (LinkSpeed, Maybe a) ->
+  -- | New state and output value
+  (Index 100, Maybe a)
+rateAdaptRxT n (linkSpeed, a)
+  | n == 0 = (n', a)
+  | otherwise = (n', Nothing)
+ where
+  n' = if n == repeatN then 0 else n + 1
+  repeatN = case linkSpeed of
+    Speed1000 -> 0
+    Speed100 -> 9
+    Speed10 -> 99
```

Comment:
I think we should only reset the count once we have received a `Just` value, i.e.:
```haskell
rateAdaptRxT n (linkSpeed, Nothing) = (n, Nothing)
rateAdaptRxT n (linkSpeed, Just a)
  | n == 0 = (n', Just a)
  | otherwise = (n', Nothing)
 where
  n' = if n == repeatN then 0 else n + 1
```
This would allow arbitrary pauses, instead of only forwarding the input every `linkSpeed` cycles.